features/steps/givens/attributes.py

```diff
@@ -74,34 +76,16 @@ def step_impl(context, inst, attribute, condition, prefix):
     elif condition == 'does not start':
         if hasattr(inst, attribute) and not str(getattr(inst, attribute, '')).startswith(prefix):
             yield ValidationOutcome(instance_id=inst, severity=OutcomeSeverity.PASSED)
-            
-
-@gherkin_ifc.step('Its final segment')
-@gherkin_ifc.step('Its final segment at depth 1')
-def step_impl(context, inst):
-    """
-    Implement ALS015
-    This is a separate function from ALB015 because ALS015 needs to return an entity_instance.
-    """
-    yield ValidationOutcome(instance_id = inst[-1], severity=OutcomeSeverity.PASSED)
-
 
-@gherkin_ifc.step('Its final IfcAlignmentSegment')
-def step_impl(context, inst):
-    """
-    Implement ALB015
-    This is a separate function from ALS015 because ALB015 needs to yield a ValidationOutcome.
-    """
-    yield ValidationOutcome(instance_id=context.instances[-1], severity=OutcomeSeverity.PASSED)
 
+@gherkin_ifc.step('Its {ff:first_or_final} element')
+@gherkin_ifc.step('Its {ff:first_or_final} element at depth 1')
+def step_impl(context, inst, ff=0, filter=None):
```

Comment:
```suggestion
def step_impl(context, inst, ff : FirstOrFinal):
```

We can probably leave out the default value. `filter` is no longer  used. Maybe add a type hint.